File: backend/src/main/java/io/ntf/api/activity/model/Activity.java
package io.ntf.api.activity.model;

import io.vavr.control.Option;
import lombok.*;
import lombok.experimental.Wither;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Duration;
import java.time.LocalDateTime;

@Document(collection = "activities")
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Wither
public class Activity {

  @Id
  private String id;

  private String userId;

  private String name;

  private LocalDateTime start;

  private LocalDateTime end;

  public Duration duration() {
    return Duration.between(start, Option.of(end).getOrElse(LocalDateTime.now()));
  }

}

File: backend/src/main/java/io/ntf/api/activity/model/Activity.kt
package io.ntf.api.activity.model

import io.vavr.control.Option
import lombok.Builder
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.Duration
import java.time.LocalDateTime

@Document(collection = "activities")
@Builder
data class Activity(@Id val id: String? = null, val userId: String, val logId: String? = null, val name: String, val start: LocalDateTime, val end: LocalDateTime? = null) {

    fun duration(): Duration {
        return Duration.between(this.start, Option.of<LocalDateTime>(this.end).getOrElse(LocalDateTime.now()))
    }

}
